src/Mod/Sketcher/App/planegcs/GCS.cpp

```diff
@@ -1326,13 +1326,63 @@ int System::addConstraintInternalAlignmentParabolaFocus(Parabola &e, Point &p1,
     return addConstraintEqual(e.focus1.y, p1.y, tagId, driving, Constraint::Alignment::InternalAlignment);
 }
 
-int System::addConstraintInternalAlignmentBSplineControlPoint(BSpline &b, Circle &c, int poleindex, int tagId, bool driving)
+int System::addConstraintInternalAlignmentBSplineControlPoint(BSpline &b, Circle &c, unsigned int poleindex, int tagId, bool driving)
 {
     addConstraintEqual(b.poles[poleindex].x, c.center.x, tagId, driving, Constraint::Alignment::InternalAlignment);
     addConstraintEqual(b.poles[poleindex].y, c.center.y, tagId, driving, Constraint::Alignment::InternalAlignment);
     return addConstraintEqual(b.weights[poleindex], c.rad, tagId, driving, Constraint::Alignment::InternalAlignment);
 }
 
+int System::addConstraintInternalAlignmentKnotPoint(BSpline &b, Point &p, unsigned int knotindex, int tagId, bool driving)
+{
+    if (b.periodic && knotindex == 0) {
+        // This is done here since knotpoints themselves aren't stored
+        addConstraintP2PCoincident(p, b.start, tagId, driving);
+        addConstraintP2PCoincident(p, b.end, tagId, driving);
+    }
+
+    size_t numpoles = b.degree - b.mult[knotindex] + 1;
+    if (numpoles == 0)
+        numpoles = 1;
+
+    size_t startpole = 0;
+    std::vector<double *> pvec;
+    pvec.push_back(p.x);
+
+    std::vector<double> weights(numpoles, 1.0 / numpoles);
+
+    // TODO: Adjust for periodic knot
+    for (size_t j = 1; j <= knotindex; ++j)
+        startpole += b.mult[j];
+    if (!b.periodic && startpole >= b.poles.size())
+        startpole = b.poles.size() - 1;
+    for (size_t i = 0; i < numpoles; ++i) {
+        pvec.push_back(b.poles[(startpole + i) % b.poles.size()].x);
+        // FIXME: `getLinCombFactor` seg-faults for the last knot so this safeguard exists
```

Comment:
I might have fixed this issue in #7607. This was causing a headache when I coded this part so maybe I didn't dare poke it further. The commits leading to this are probably squashed even in the [backup](https://github.com/AjinkyaDahale/FreeCAD/tree/sketcher-constraint-knots-backup) so this needs fresh testing.

Reply:
Just tried removing the condition and no segfaults now, somehow. However, this causes some nan values which eventually go into the infinite loop in diagnosis (I suspect the same one plaguing tangent-at-knot). I do not expect this happen with this check.

The reason the nan comes up is because of the `k` parameter passed on to `getLinCombFactor`. For the last knot, this translates to the very last `flattenedknots` interval, which for a non-periodic b-spline is [1, 1], a degenerate case. The cleanest way to handle this currently is to keep this condition. Otherwise, the appropriate interval needs to be found for one specific case. It also has the effect on performance since we don't need to run the de Boor algorithm just to get 1.0 or 0.5.

Reply:
I just modified the code slightly for cleanliness and added the above explanation there. Resolving this issue. Please reopen/complain if this is not satisfactory.